### PR TITLE
cheri_compartment: Warn if return type is void or return value is unused

### LINKED ISSUE #34: Warn if compartment call returns void
It would be good if `clang` would warn if the return type of a compartment call is `void`. Compartment calls can fail unexpectedly and return `-1` (e.g. due to trusted stack exhaustion). Callers should be aware of and check for this, therefore it is never appropriate for a compartment call to return `void`. For the same reason we should also consider automatically adding the `nodiscard` attribute to compartment calls.

### PR DESCRIPTION
**Pull request by v01dXYZ**
Fixes #34.

### CODE REVIEW

**Comment by davidchisnall** [comment]
Thanks for doing this. We are in the middle of rebasing our changes on a newer LLVM, so I’d like to hold off merging until that’s done.

The warning would be better if it is explicit about why it’s a bad idea, rather than saying that it’s unsupported. I’d recommend something like:

void return on a cross-compartment call makes it impossible for callers to detect failure.

A fixit hint that changed the type to int and changed return statements to return 0 (and adds one at the end of the function) would be ideal.

**Comment by v01dXYZ** [comment]
Thank you. I'll look at FixItHint.

**Comment by v01dXYZ** [comment]
*This comment is for documentation.*

* note diagnosis are not meant to be emitted alone but in conjunction with a warning/error.  If you do it nonetheless, it is very likely it would be ignored as the last non-note diagnosis could have been an ignored warning (and there are a lot of those). 
* emiting a diagnosis at a given location with a fixit at another line (location?) doesn't work. In this case, use warning + note. If you do it nonetheless, the fixit will be ignored.
* `~FuncDec::getFunctionTypeLoc` fails when there are attributes. The TypeSourceInfo has a TypeLoc with the following structure: AttributedType, FunctionTypeProto.

**Comment by v01dXYZ** [comment]
I need to adapt an existing function, I will reimplement it where it's needed but I also opened an issue to know if it's acceptable to directly modify the function:  llvm/llvm-project#118290.

**Comment by v01dXYZ** [comment]
@davidchisnall

Here the following error messages:

* void return type:
  * Warning: `void return on a cross compartment call makes it impossible for callers to detect failures`
  * Note: `Replace void return type with int` + fix-it hint replacement
* `return;`, no return:  
  * Warning: `cross-compartment calls that always succeed should return 0 instead` + fix-it hint insertion

Do you agree with the diagnosis levels (warn, note, ...) ? Don't forget to comment on capitalisation or ending the messages with dot (or not). Just asking to not turn this PR into a Nicaea council.

I plan to add this (corrected) description of the messages to the issue description.

**Comment by davidchisnall** [comment]
That sounds good to me as a structure.  @rmn30 / @nwf, any opinions on the message text?

**Comment by rmn30** [comment]
Sounds good!

**Comment by nwf** [comment]
Sounds good to me (and it's a useful diagnostic to have; thanks!)

**Comment by v01dXYZ** [comment]
This PR is almost finished but there is some parts that depend on modifying the behavior of some core clang class (FunctionDecl::getReturnTypeLoc). I'm working on a patch for trunk llvm and waiting for review. This modification is needed for the FixItHint to replace void by int. I think it's going to take 2 more weeks.

**Comment by davidchisnall** [comment]
I don't mind carrying the `FunctionDecl::getReturnTypeLoc` patch here while it's being upstreamed.  We can probably move to LLVM 17 this week, and then merge this.

**Comment by resistor** [comment]
You're probably going to need to cherry-pick this onto the new llvm-17 based branch.

**Comment by v01dXYZ** [comment]
The CI timed out. Is there any way to relaunch it without pushing ?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 1]
```diff
@@ -2336,6 +2336,13 @@ def note_in_reference_temporary_list_initializer : Note<
   "list-initialize this reference">;
 def note_var_fixit_add_initialization : Note<
   "initialize the variable %0 to silence this warning">;
+def warn_cheri_compartment_void_return_type : Warning <
+  "void return on a cross-compartment call make it impossible for callers to detect failure">,
```
Comment (resistor) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td]: "makes it impossible"

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 2]
```diff
@@ -2336,6 +2336,13 @@ def note_in_reference_temporary_list_initializer : Note<
...
+  InGroup<CHERICompartmentReturnVoid>;
+def note_cheri_compartment_void_return_type : Note<"replace void return type with int">;
+def warn_cheri_compartment_return_void_or_falloff : Warning <
+  "Cross-compartement calls that always succeed should return 0 instead">,
```
Comment (resistor) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td]: Shouldn't begin with a capital

File: clang/lib/AST/Decl.cpp [hunk 3]
```diff
@@ -3745,8 +3745,23 @@ bool FunctionDecl::doesDeclarationForceExternallyVisibleDefinition() const {
 
 FunctionTypeLoc FunctionDecl::getFunctionTypeLoc() const {
   const TypeSourceInfo *TSI = getTypeSourceInfo();
-  return TSI ? TSI->getTypeLoc().IgnoreParens().getAs<FunctionTypeLoc>()
-             : FunctionTypeLoc();
+
```
Comment (resistor) [inline clang/lib/AST/Decl.cpp]: Can you add a comment here like so:

`// CHERIOT: ...` that explains why we're carrying a divergence from upstream.

**Review by resistor — APPROVED** [review]
LGTM

**Comment by resistor** [comment]
@v01dXYZ I'm not sure if you can merge this of it you need me to do it. If the latter, please rebase it and then ping me to merge it.

**Comment by v01dXYZ** [comment]
It seems it's pretty easy to run out of credits with the Cirrus CI free plan.

**Comment by resistor** [comment]
@v01dXYZ Getting a report from @nwf that this is causing linker errors in cheriot-rtos. I'm going to revert these changes for now. I won't have time to look into the failure in detail in the next two days. If you figure out what's wrong in the meantime, go ahead and re-PR it. Otherwise I'll help figure out the problem when I'm around again.

**Comment by resistor** [comment]
@nwf reports that git bisect points to: https://github.com/cheriot-Platform/llvm-project/commit/4d8ad0bffbe890268bb076de6b5b2f526324bc87

**Comment by resistor** [comment]
The full link error is:

```
[ 99%]: linking firmware build/cheriot/cheriot/release/test-suite
/home/ubuntu/llvm-project-1/build/install/bin/ld.lld -n --script=build/test-suite-firmware.ldscript --relax -o build/cheriot/cheriot/release/test-suite --compartment-report=build/cheriot/cheriot/release/test-suite.json build/.objs/cheriot.loader/cheriot/cheriot/release/__/sdk/core/loader/boot.S.o build/.objs/cheriot.loader/cheriot/cheriot/release/__/sdk/core/loader/boot.cc.o build/.objs/cheriot.switcher/cheriot/cheriot/release/__/sdk/core/switcher/entry.S.o build/cheriot/cheriot/release/allocator_test.compartment build/cheriot/cheriot/release/atomic1.library build/cheriot/cheriot/release/atomic2.library build/cheriot/cheriot/release/atomic4.library build/cheriot/cheriot/release/atomic8.library build/cheriot/cheriot/release/ccompile_test.compartment build/cheriot/cheriot/release/check_pointer_test.compartment build/cheriot/cheriot/release/cheriot.allocator.compartment build/cheriot/cheriot/release/compartment_helpers.library build/cheriot/cheriot/release/locks.library build/cheriot/cheriot/release/cheriot.software_revoker.compartment build/cheriot/cheriot/release/cheriot.token_library.library build/cheriot/cheriot/release/compartment_calls_inner.compartment build/cheriot/cheriot/release/compartment_calls_test.compartment build/cheriot/cheriot/release/crash_recovery_inner.compartment build/cheriot/cheriot/release/crash_recovery_outer.compartment build/cheriot/cheriot/release/crash_recovery_test.compartment build/cheriot/cheriot/release/crt.library build/cheriot/cheriot/release/cxxrt.library build/cheriot/cheriot/release/debug.library build/cheriot/cheriot/release/debug_test.compartment build/cheriot/cheriot/release/event_group.library build/cheriot/cheriot/release/eventgroup_test.compartment build/cheriot/cheriot/release/freestanding.library build/cheriot/cheriot/release/futex_test.compartment build/cheriot/cheriot/release/list_test.compartment build/cheriot/cheriot/release/locks_test.compartment build/cheriot/cheriot/release/message_queue.compartment build/cheriot/cheriot/release/message_queue_library.library build/cheriot/cheriot/release/misc_test.compartment build/cheriot/cheriot/release/mmio_test.compartment build/cheriot/cheriot/release/multiwaiter_test.compartment build/cheriot/cheriot/release/queue_test.compartment build/cheriot/cheriot/release/stack_integrity_thread.compartment build/cheriot/cheriot/release/stack_test.compartment build/cheriot/cheriot/release/static_sealing_inner.compartment build/cheriot/cheriot/release/static_sealing_test.compartment build/cheriot/cheriot/release/stdio.library build/cheriot/cheriot/release/string.library build/cheriot/cheriot/release/stdio_test.compartment build/cheriot/cheriot/release/test-suite.scheduler.compartment build/cheriot/cheriot/release/test_runner.compartment build/cheriot/cheriot/release/thread_pool.compartment build/cheriot/cheriot/release/thread_pool_test.compartment build/cheriot/cheriot/release/unwind_cleanup_test.compartment
ld.lld: error: undefined symbol: heap_allocate
>>> referenced by global_constructors.hh:49 (/workspaces/cheriot-rtos/sdk/include/c++-config/global_constructors.hh:49)
>>>               build/cheriot/cheriot/release/allocator_test.compartment:()
>>> referenced by global_constructors.hh:29 (/workspaces/cheriot-rtos/sdk/include/c++-config/global_constructors.hh:29)
>>>               build/cheriot/cheriot/release/allocator_test.compartment:()
>>> referenced by global_constructors.hh:0 (/workspaces/cheriot-rtos/sdk/include/c++-config/global_constructors.hh:0)
>>>               build/cheriot/cheriot/release/allocator_test.compartment:()
>>> referenced 17 more times

ld.lld: error: undefined symbol: heap_allocate_array
>>> referenced by __cpp_allocator-test.cc.cc
>>>               build/cheriot/cheriot/release/allocator_test.compartment:()
>>> referenced by __cpp_allocator-test.cc.cc
>>>               build/cheriot/cheriot/release/allocator_test.compartment:()

ld.lld: error: undefined symbol: revoker_epoch_get()
>>> referenced by main.cc:931 (../sdk/core/allocator/main.cc:931)
>>>               build/cheriot/cheriot/release/cheriot.allocator.compartment:()

ld.lld: error: undefined symbol: __library_export_libcalls__Z6memcpyPvPKvj
>>> referenced by __cpp_misc-test.cc.cc
>>>               build/cheriot/cheriot/release/misc_test.compartment:()

ld.lld: error: undefined symbol: __export__heap_allocate
>>> referenced by __cpp_queue.cc.cc
>>>               build/cheriot/cheriot/release/message_queue_library.library:()
>>> referenced by __cpp_event_group.cc.cc
>>>               build/cheriot/cheriot/release/event_group.library:()
```

The `memcpy` part is due to another commit that I'm also backing out, but the rest are due to this PR.

**Comment by v01dXYZ** [comment]
I suspect a strange behavior from xmake. My reasoning: compilation should fail as there is the Werror flag set. There are some bits of code that have a void return type or ignore the return value of a compartment call. I wonder also if it creates a empty file instead of no file, this file is then ignored by the linker.

**Comment by resistor** [comment]
I made some time to look into it, and I think I found it. I'll push a PR for you to review soon.

**Comment by v01dXYZ** [comment]
At least it forced me to compile cheriot-rtos tests. I'll add a line in my local cirrus.yml for doing that since it only requires installing xmake which is debian/Ubuntu available

**Comment by rmn30** [comment]
>  xmake which is debian/Ubuntu available

Be careful as there we were hit by some issues in various versions of xmake and the version available in debian / ubuntu repos may not work correctly. At present the devcontainer Dockerfile is installing a version from a [ppa](https://launchpad.net/~xmake-io/+archive/ubuntu/xmake) but that may only work on Ubuntu.

**Pull request merged**